### DISCUSSION TITLE
Async method passing self

### DISCUSSION THREAD

**Discussion by perseoGI** [rwf2/Rocket#2556] [Questions] [answered]
I there, I've been strugling with `rocket_sync_db_pools` & `diesel_postgres_pool` and async code.
```toml
[dependencies.rocket_sync_db_pools]
version = "0.1.0-rc.3"
features = ["diesel_postgres_pool"]
[dependencies.diesel]
version = "2.1.0"
features = ["postgres", "r2d2", "chrono"]
```
Imagine following code:

```rust
#[derive(Debug, Identifiable, Queryable, Associations, Deserialize, Serialize)]
#[diesel(table_name = images)]
#[diesel(belongs_to(Part, foreign_key = part_id))]
pub struct Image {
    pub id: Id,
    pub part_id: Id,
    pub path: String,
}

#[derive(Identifiable, Queryable, Serialize)]
pub struct Part {
    pub id: Id,
    ...
}
#[derive(Queryable, Debug, Serialize)]
pub struct PartDTO {
    pub id: Id,
    pub images: Vec<(Id, String)>,
}
```

Lets say I want to use the `belonging_to` method to query all images belonging to a part this way:

```rust
pub async fn to_dto(&self, conn: &DbConn) -> PartDTO {
    let images: Vec<(Id, String)> = conn.run(move |c| {
        Image::belonging_to(self)
            .get_results::<Image>(c)
    })
    .await
    .unwrap()
    .into_iter()
    .map(|image| {
        (
            image.id,
            Path::new(&image.path)
                .file_stem()
                .unwrap()
                .to_os_string()
                .to_str()
                .unwrap()
                .into(),
        )
    })
    .collect();

  PartDTO {id: self.id, images}
} 
```

So that, in any services (such as create, get_one, get_all, etc...) I could end up calling thes `to_dto` method, eg of create:
```rust
pub async fn create(part: PartInputDTO, conn: &DbConn) -> QueryResult<PartDTO> {
       let created_part = conn
        .run(move |c| {
            diesel::insert_into(parts)
                .values(&part_input)
                .get_result::<Part>(c)
        })
        .await;

    Ok(created_part.unwrap().to_dto(conn).await)
}
```

Sadly, I get the following compile time error:

```
self` is a reference that is only valid in the method body let's call the lifetime of this reference `'1`
`self` escapes the method body here (conn.run)  argument requires that `'1` must outlive `'static`
```
I've tried making the `self` parameter `&'static Self`, using `std::rc::Rc pointers` to clone the self object, using `async-trait` lib, but all of those solutions seems to fail. 

I guess I could avoid using `belonging_to` function and use a standard query, but this code used to work in diesel v1 and I would love to keep it using and understand how to fix this.

Thanks in advance!

**Comment by SergioBenitez** [comment, marked as answer]
In Rocket 0.4, which wasn't `async`, running such a query as part of a handler would stall that thread while the database responded to the query. Because execution could not continue until the database responded, it was okay to reference any local data part of the query.

In Rocket v0.5, the query is run in the background and does not stall the handler's thread. As a result, if a reference to local data were allowed in the query, that reference could outlive the stack, which it points to, which would result in a dangling pointer. As a result, the `.run()` method requires that any references in the passed-in closure live at least as long as `'static`, i.e, the program itself.

Now, using `.await` on `run()` effectively blocks the thread (in a cooperative manner, now) as before, but we're still not allowed to hold references in `run()` because 1) the `.await` need-not be called, and more importantly, 2) `diesel` itself is synchronous and thus uses of it _must_ be run in a background thread or else the `async` executor (tokio) can be blocked, leading to very poor performance. If `diesel` itself were `async`, none of this would be a problem, and your code would both look nicer and work more as expected. But that isn't the case today. You could switch to an `async` database driver like `sqlx`, of course.

Knowing this, and if we continue to use `diesel`, we need to abide by the restrictions. The problem in your code is here:

> ```rust
> pub async fn to_dto(&self, conn: &DbConn) -> PartDTO {
>     let images: Vec<(Id, String)> = conn.run(move |c| {
>         Image::belonging_to(self)
>             .get_results::<Image>(c)
>     })
> ```

The reference `self` does not outlive the program and never will. No amount of lifetime annotations will help. `Part` here needs to be `'static`. Here's one way to fix it:

```diff
- #[derive(Debug, Identifiable, Queryable, Associations, Deserialize, Serialize)]
+ #[derive(Debug, Clone, Identifiable, Queryable, Associations, Deserialize, Serialize)]
#[diesel(table_name = images)]
#[diesel(belongs_to(Part, foreign_key = part_id))]
pub struct Image {
    pub id: Id,
    pub part_id: Id,
    pub path: String,
}

pub async fn to_dto(&self, conn: &DbConn) -> PartDTO {
+    let part = self.clone();
    let images: Vec<(Id, String)> = conn.run(move |c| {
-        Image::belonging_to(self)
+        Image::belonging_to(&part)
            .get_results::<Image>(c)
    })
```

**Reply by perseoGI** [reply]
Thank you very much for that incredible explanation. That worked like a charm!

I'll keep using `diesel` for now as it works fine for me, but I could take a look into `sqlx` in the future! 
Thank you again :rocket: 

This solves [this](https://github.com/diesel-rs/diesel/discussions/3649) question

**Reply by weiznich** [reply]
> diesel itself is synchronous and thus uses of it must be run in a background thread or else the async executor (tokio) can be blocked, leading to very poor performance. If diesel itself were async, none of this would be a problem, and your code would both look nicer and work more as expected. But that isn't the case today. You could switch to an async database driver like sqlx, of course.

This is only partially correct. It's correct in so far that `diesel` (as in literally the crate named `diesel`) offers a `sync` database implementation only. It's incorrect in so far that `diesel` support async database drives (via the [`diesel-async`](https://lib.rs/crates/diesel-async) crate.). So complaining here that this is a problem because `diesel` does not support async is not true!
(What might be missing is an integration for `diesel-async` on rockets side, but that's then arguably a rocket issue, not an diesel issue!)
